lib/Dialect/ESI/runtime/cpp/lib/backends/Cosim.cpp

```diff
@@ -80,147 +108,212 @@ CosimAccelerator::CosimAccelerator(Context &ctxt, string hostname,
   }
-  virtual void disconnect() override {
-    if (ep)
-      ep->returnForUse();
+
+  /// Send a write message to the server.
+  void write(const MessageData &data) override {
+    ClientContext context;
+    AddressedMessage msg;
+    msg.set_channel_name(name);
+    msg.mutable_message()->set_data(data.getBytes(), data.getSize());
+    VoidMessage response;
+    grpc::Status sendStatus = rpcClient->SendToServer(&context, msg, &response);
+    if (!sendStatus.ok())
+      throw runtime_error("Failed to write to channel '" + name +
+                          "': " + sendStatus.error_message() +
+                          ". Details: " + sendStatus.error_details());
   }
-  virtual void write(const MessageData &) override;
 
 protected:
-  esi::cosim::Endpoint *ep;
+  ChannelServer::Stub *rpcClient;
+  /// The channel description as provided by the server.
+  ChannelDesc desc;
+  /// The name of the channel from the manifest.
   string name;
 };
 } // namespace
 
-void WriteCosimChannelPort::write(const MessageData &data) {
-  ep->pushMessageToSim(make_unique<esi::MessageData>(data));
-}
-
 namespace {
-class ReadCosimChannelPort : public ReadChannelPort {
+/// Cosim client implementation of a read channel port. Since gRPC read protocol
+/// streams messages back, this implementation is quite complex.
+class ReadCosimChannelPort
+    : public ReadChannelPort,
+      public grpc::ClientReadReactor<esi::cosim::Message> {
 public:
-  ReadCosimChannelPort(esi::cosim::Endpoint *ep, const Type *type, string name)
-      : ReadChannelPort(type), ep(ep), name(name) {}
-  virtual ~ReadCosimChannelPort() = default;
+  ReadCosimChannelPort(ChannelServer::Stub *rpcClient, const ChannelDesc &desc,
+                       const Type *type, string name)
+      : ReadChannelPort(type), rpcClient(rpcClient), desc(desc), name(name),
+        context(nullptr) {}
+  virtual ~ReadCosimChannelPort() { disconnect(); }
 
-  // TODO: Replace this with a request to connect to the capnp thread.
   virtual void connect() override {
-    if (!ep)
-      throw runtime_error("Could not find channel '" + name +
-                          "' in cosimulation");
-    if (ep->getRecvTypeId() == "")
-      throw runtime_error("Channel '" + name + "' is not a read channel");
-    if (ep->getRecvTypeId() != getType()->getID())
+    // Sanity checking.
+    ReadChannelPort::connect();
+    if (desc.type() != getType()->getID())
       throw runtime_error("Channel '" + name + "' has wrong type. Expected " +
-                          getType()->getID() + ", got " + ep->getRecvTypeId());
-    ep->setInUse();
+                          getType()->getID() + ", got " + desc.type());
+    if (desc.dir() != ChannelDesc::Direction::ChannelDesc_Direction_TO_CLIENT)
+      throw runtime_error("Channel '" + name + "' is not a to server channel");
+    assert(desc.name() == name);
+
+    // Initiate a stream of messages from the server.
+    context = new ClientContext();
```

Comment:
as above - is there a particular good argument for having raw pointers here? Else, smart pointers should be the default.